torch/_inductor/codegen/cuda/cuda_kernel.py

```diff
@@ -265,21 +273,283 @@ def row_or_column_stride(self, node: IRNode, default_value: int = 0) -> str:
+        """
+
+        Check if the given node can be fused with the epilogue. At the moment, Kernels
+        support fusion with Pointwise operations, wrapped in (named) ComputedBuffer nodes.
+
+        Args:
+            node: The IRNode to be check if it can be fused with the epilogue.
+
+        Returns:
+        - bool: True if the given node can be fused with the epilogue, False otherwise.
+
+        """
+        if not self.template.can_fuse_epilogue:
+            # The used GEMM op does not support fusing epilogues
+            return False
+
+        if not isinstance(node, ComputedBuffer):
+            return False
+        if not isinstance(node.data, Pointwise):
+            return False
+
+        # We can fuse a Pointwise op that depends on the last fused epilogue node
+        # if any. If there is no epilogue node yet, it needs to depend on the template
+        # node
+        node_name = node.name if node.name is not None else node.data.name  # type: ignore[attr-defined]
+        if node_name is None:
+            return False
+
+        if len(self._epilogue_nodes) == 0:
+            if self.name not in node.get_read_names():
+                return False
+        else:
+            last_epilogue_node = cast(ComputedBuffer, self._epilogue_nodes[-1])
+            last_epilogue_name = (
+                last_epilogue_node.name
+                if last_epilogue_node.name is not None
+                else last_epilogue_node.data.name  # type: ignore[attr-defined]
+            )
+            if last_epilogue_name not in node.get_read_names():
+                return False
+        if node.layout != self.layout:
+            return False
+        try:
+            from torch._inductor.codegen.cuda.cutlass_epilogue_gen import (
+                CutlassEVTEpilogueArgumentFormatter,
+                CutlassEVTEpilogueTypeFormatter,
+            )
+
+            CutlassEVTEpilogueTypeFormatter.ir_to_evt_string(
+                cast(str, self.name), "anything", [node]
+            )
+            CutlassEVTEpilogueArgumentFormatter.ir_to_evt_argument_string(
+                cast(str, self.name), [node]
+            )
+        except NotImplementedError as e:
+            not_implemented_op = str(e)
+            if not_implemented_op.startswith("_op_"):
+                not_implemented_op = not_implemented_op[4:]
+                log.warning(
+                    f"Cannot fuse epilogue node {node} into {self.name}, likely due to unsupported operation: {not_implemented_op}"  # noqa: G004, B950
+                )
+                return False
+            else:
+                # Likely due to unsupported dtype.
+                log.warning(
+                    f"Cannot fuse epilogue node {node} into {self.name}. Reason: {not_implemented_op}"  # noqa: G004, B950
+                )
+                return False
+        return True
+
+    def create_fused_buffer(self, node, scheduler: Scheduler) -> "CUDATemplateBuffer":
```

Comment:
I need to take a second pass of the PR, but at first glance this looks a bit weird. 
`CUDATemplateBuffer` belongs to IR definition, which should be finalized at the lowering stage.
This function is called at scheduling time. It seems to me that we should not reuse an IR concept here for codegen.

Reply:
Yes, it's a bit tricky to get that right and I'm also not really satisfied with this. 

So, let me list the changes that would be necessary (at least) to decouple this:

 * The entire "can_fuse" decision would have to be pushed out of CUDATemplateBuffer into a separate function (accepting both the CUDATemplateBuffer and all epilogue nodes ) that is then called by CUDASchedulerNode and FusedCUDASchedulerNode
 * The CUDATemplateKernel.call_kernel method would have to accept additional epilogue nodes and fuse lists of  inputs itself to generate the call signature.
 * The CUDATemplateBuffer would have to pass around the epilogue nodes through many method calls (starting with render), where at the moment it can pass these as attribute on self.

So, it's doable but a bit of work. I can either revise this PR or push this into phase 2.

Reply:
I'd prefer decoupling logics for ir and codegen. It's much easier to reason about codes.

Reply:
I implemented these changes in the latest update to this PR..